config/v1/types.go

```diff
@@ -80,6 +81,123 @@ type ConfigMapReference struct {
 	Key string `json:"filename,omitempty"`
 }
 
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Build holds cluster-wide information on how to handle builds. The canonical name is `cluster`
+type Build struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// Spec holds user-settable values for the build controller configuration
+	// +optional
+	Spec BuildSpec `json:"spec,omitempty"`
+}
+
+type BuildSpec struct {
+	// AdditionalTrustedCA is a reference to a ConfigMap containing additional CAs that
+	// should be trusted for image pushes and pulls during builds.
+	// +optional
+	AdditionalTrustedCA ConfigMapReference `json:"additionalTrustedCA,omitempty"`
+	// BuildDefaults controls the default information for Builds
+	// +optional
+	BuildDefaults *BuildDefaults `json:"buildDefaults,omitempty"`
+	// BuildOverrides controls override settings for builds
+	// +optional
+	BuildOverrides *BuildOverrides `json:"buildOverrides,omitempty"`
+}
+
+type BuildDefaults struct {
+	// GitHTTPProxy is the location of the HTTPProxy for Git source
+	// +optional
+	GitHTTPProxy string `json:"gitHTTPProxy,omitempty"`
+
+	// GitHTTPSProxy is the location of the HTTPSProxy for Git source
+	// +optional
+	GitHTTPSProxy string `json:"gitHTTPSProxy,omitempty"`
+
+	// GitNoProxy is the list of domains for which the proxy should not be used
+	// +optional
+	GitNoProxy string `json:"gitNoProxy,omitempty"`
+
+	// Env is a set of default environment variables that will be applied to the
+	// build if the specified variables do not exist on the build
+	// +optional
+	Env []corev1.EnvVar `json:"env,omitempty"`
+
+	// SourceStrategyDefaults are default values that apply to builds using the
+	// source strategy.
+	// +optional
+	SourceStrategyDefaults *SourceStrategyDefaults `json:"sourceStrategyDefaults,omitempty"`
+
+	// ImageLabels is a list of docker labels that are applied to the resulting image.
+	// User can override a default label by providing a label with the same name in their
+	// Build/BuildConfig.
+	// +optional
+	ImageLabels []ImageLabel `json:"imageLabels,omitempty"`
+
+	// NodeSelector is a selector which must be true for the build pod to fit on a node
+	// +optional
+	NodeSelector *metav1.LabelSelector `json:"nodeSelector,omitempty"`
+
+	// Annotations are annotations that will be added to the build pod
+	// +optional
+	Annotations map[string]string `json:"annotations,omitempty"`
+
+	// Resources defines resource requirements to execute the build.
+	// +optional
+	Resources corev1.ResourceRequirements `json:"resources,omitempty"`
+}
+
+type SourceStrategyDefaults struct {
+
+	// Incremental indicates if s2i build strategies should perform an incremental
+	// build or not
+	// +optional
+	Incremental *bool `json:"incremental,omitempty"`
```

Comment:
The problem is this is a defaulter.   We need to distinguish between "user did not want to specify a default for this field" and "the user has requested the default value be X"

if we make the field non-optional and default it to 'false' then we have to treat it as if every single cluster everywhere has an admin who said "i want the default to be incremental=false".  

While that's not strictly a problem today (there is no difference between a cluster with a cluster-wide default of incremental=false, and one with no default, because the builds themselves default to incremental=false anyway), it does cause us to lose user intent and may make migration problematic in the future, if for example we wanted to move to a default incremental behavior of true at the build level.

basically it's a second order issue, i think.

nil -> don't touch the user's build (so either you get what the user asked for, or whatever the default build behavior is, which today is false but in the future could be true, theoretically)
false -> set the value to false in the user's build if the user has not set a value in their build
true -> set the value to true in the user's build if the user has not set a value in their build

Reply:
> nil -> don't touch the user's build (so either you get what the user asked for, or whatever the default build behavior is, which today is false but in the future could be true, theoretically)
> false -> set the value to false in the user's build if the user has not set a value in their build
> true -> set the value to true in the user's build if the user has not set a value in their build


That first one is really hard to follow, but I can see the theoretical value I supposed.  Defaulting bools is awful to both think about an explain.